tests/plotting/bokeh/testserver.py

```diff
@@ -0,0 +1,157 @@
+        bokeh_renderer.last_plot = None
+        Callback._callbacks = {}
+
+    def test_render_server_doc_element(self):
+        obj = Curve([])
+        doc = bokeh_renderer.server_doc(obj)
+        self.assertIs(doc, curdoc())
+        self.assertIs(bokeh_renderer.last_plot.document, curdoc())
+
+    def test_render_explicit_server_doc_element(self):
+        obj = Curve([])
+        doc = Document()
+        server_doc = bokeh_renderer.server_doc(obj, doc)
+        self.assertIs(server_doc, doc)
+        self.assertIs(bokeh_renderer.last_plot.document, doc)
+
+    def test_set_up_linked_change_stream_on_server_doc(self):
+        obj = Curve([])
+        stream = RangeXY(source=obj)
+        server_doc = bokeh_renderer.server_doc(obj)
+        self.assertEqual(len(bokeh_renderer.last_plot.callbacks), 1)
+        cb = bokeh_renderer.last_plot.callbacks[0]
+        self.assertIsInstance(cb, RangeXYCallback)
+        self.assertEqual(cb.streams, [stream])
+        x_range = bokeh_renderer.last_plot.handles['x_range']
+        self.assertIn(cb.on_change, x_range._callbacks['start']) 
+        self.assertIn(cb.on_change, x_range._callbacks['end'])
+        y_range = bokeh_renderer.last_plot.handles['y_range']
+        self.assertIn(cb.on_change, y_range._callbacks['start']) 
+        self.assertIn(cb.on_change, y_range._callbacks['end'])
+
+    def test_set_up_linked_event_stream_on_server_doc(self):
+        obj = Curve([])
+        stream = PlotReset(source=obj)
+        server_doc = bokeh_renderer.server_doc(obj)
+        cb = bokeh_renderer.last_plot.callbacks[0]
+        self.assertIsInstance(cb, ResetCallback)
+        self.assertEqual(cb.streams, [stream])
+        plot = bokeh_renderer.last_plot.state
+        self.assertIn(cb.on_event, plot._event_callbacks['reset']) 
+         
+    
+
+class TestBokehServerRun(ComparisonTestCase):
+
+    def setUp(self):
+        self.previous_backend = Store.current_backend
+        if not bokeh_renderer:
+            raise SkipTest("Bokeh required to test plot instantiation")
+        Store.current_backend = 'bokeh'
+
+    def tearDown(self):
+        Store.current_backend = self.previous_backend
+        Callback._callbacks = {}
+
+    def test_launch_simple_server(self):
+        obj = Curve([])
+        global launched
+        launched = False
+        def modify_doc(doc):
+            bokeh_renderer(obj, doc=doc)
+            global launched
+            launched = True
+            server.stop()
+        handler = FunctionHandler(modify_doc)
+        app = Application(handler)
+        server = Server({'/': app}, port=0)
+        server.start()
+        url = "http://localhost:" + str(server.port) + "/"
+        pull_session(session_id='Test', url=url, io_loop=server.io_loop)
+        self.assertTrue(launched)
+
+    def test_launch_server_with_stream(self):
+        obj = Curve([])
+        stream = RangeXY(source=obj)
+
+        global launched
```

Comment:
If you need a global, maybe it should be something like a dict so you can say that it was launched for a particular unit test? Or do you not see any need for this sort of test anywhere else?

Reply:
Well I specifically set it to False each time before the test, so I don't think that's needed.

Reply:
As an alternative to using a global, something like this should work:

```
launched = []
def modify_doc(..):
   ...
   launched.append(True)
...
self.assertEquals(len(launched)==1)
```

Reply:
Happy to change it to that.